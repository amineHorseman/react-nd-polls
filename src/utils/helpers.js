import { FaCircleUser } from "react-icons/fa6"

export function displayAvatar(avatarURL) {
    // Display user avatar, or a profile icon if URL is null
    return (
        <span className="avatar">
        { avatarURL === null ? 
            <FaCircleUser /> : 
            <img src={avatarURL} className="rounded-circle" alt="User avatar"/>
        }
        </span>
    );
}

export function displayElapsedTime(timestamp) {
    // Display elapsed time since 'timestamp' in a readable text format
    const elapsedTimeString = getElapsedTime(timestamp);
    return <span className="text-muted">{ elapsedTimeString }</span>;
}

export function getElapsedTime(timestamp) {
    // Compute the elapsed time since 'timestamp', and returns it as a readable string
    const elapsedTimestamp = Date.now() - timestamp; 
    const minutes = Math.floor(elapsedTimestamp / (1000 * 60));
    const hours = Math.floor(minutes / 60);
    const days = Math.floor(hours / 24);
    const months = Math.floor(days / 30);
    const years = Math.floor(months / 12);

    if (years > 1)
        return `${years} year(s) ago`
    else if (months > 1)
            return `${months} days ago`
    else if (days > 1)
        return `${days} days ago`
    else if (hours > 1)
        return `${hours} hours ago`
    else if (minutes > 1)
        return `${minutes} minutes ago`
    else
        return 'less than a minute ago'
}

export const handleFormChange = (e, field, setFormState) => {
    setFormState(formState => ({...formState, 
        [field]: e.target.value,
        error: ''}));
}
