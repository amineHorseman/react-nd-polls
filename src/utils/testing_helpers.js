import { fetchQuestions } from "../features/questions/questionsSlice";
import { fetchUsers } from "../features/users/usersSlice";
import { loginUser } from "../features/auth/authSlice";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import { Provider } from "react-redux";
import store from '../app/store';

export function wrapAndRenderComponent(component) {
    return render(
        <BrowserRouter>
            <Provider store={store}>
                {component}
            </Provider>
        </BrowserRouter>
    );
}

export async function loadInitialDataForTesting() {
    await store.dispatch(fetchUsers());
    await store.dispatch(fetchQuestions());
}

export async function loginUserForTesting(username) {
    await store.dispatch(loginUser(username));
}
