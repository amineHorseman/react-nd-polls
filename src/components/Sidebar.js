import { selectUser } from "../features/users/usersSlice";
import { selectAuthedUserId } from "../features/auth/authSlice";
import { useSelector } from "react-redux";
import { displayUserAvatar } from "../utils/helpers";

const Sidebar = () => {
    const username = useSelector(selectAuthedUserId);
    const user = useSelector((state) => selectUser(state, username));

    return <div className="sidebar">
        <div className="card">
            <div className="card-body text-center">
                <div className="mb-2">
                    { displayUserAvatar(user["avatarURL"]) }   
                </div>
                <h5 className="card-title mb-0">{username}</h5>
                <p className="card-text text-muted">{user["name"]}</p>
                <hr />
                <div className="row text-center">
                    <div className="col-6">
                        <h6 className="mb-0">{user["questions"].length}</h6>
                        <small className="text-muted">Questions</small>
                    </div>
                    <div className="col-6">
                        <h6 className="mb-0">{Object.keys(user["answers"]).length}</h6>
                        <small className="text-muted">Answers</small>
                    </div>
                </div>
            </div>
        </div>
    </div>;
};

export default Sidebar;