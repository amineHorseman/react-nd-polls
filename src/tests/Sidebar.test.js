import '@testing-library/jest-dom';
import { loadInitialDataForTesting } from "../utils/testing_helpers";
import { wrapAndRenderComponent } from '../utils/testing_helpers';
import { loginUserForTesting } from "../utils/testing_helpers";
import Sidebar from "../components/Sidebar";

describe("Test for Sidebar component", () => {

    test("Component matches snapshot with user mtsamis", async () => {
        await loadInitialDataForTesting();
        await loginUserForTesting('mtsamis');
        const view = wrapAndRenderComponent(<Sidebar />);
        expect(view).toMatchSnapshot();
    });
});
