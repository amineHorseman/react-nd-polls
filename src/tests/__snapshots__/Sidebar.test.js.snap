// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Sidebar component Component matches snapshot with user mtsamis 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sidebar"
      >
        <div
          class="card"
        >
          <div
            class="card-body text-center"
          >
            <div
              class="mb-2"
            >
              <span
                class="avatar"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M399 384.2C376.9 345.8 335.4 320 288 320l-64 0c-47.4 0-88.9 25.8-111 64.2c35.2 39.2 86.2 63.8 143 63.8s107.8-24.7 143-63.8zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zm256 16a72 72 0 1 0 0-144 72 72 0 1 0 0 144z"
                  />
                </svg>
              </span>
            </div>
            <h5
              class="card-title mb-0"
            >
              mtsamis
            </h5>
            <p
              class="card-text text-muted"
            >
              Mike Tsamis
            </p>
            <hr />
            <div
              class="row text-center"
            >
              <div
                class="col-6"
              >
                <h6
                  class="mb-0"
                >
                  5
                </h6>
                <small
                  class="text-muted"
                >
                  Total Score
                </small>
              </div>
              <div
                class="col-6"
              >
                <h6
                  class="mb-0"
                >
                  2
                </h6>
                <small
                  class="text-muted"
                >
                  Ranking
                </small>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sidebar"
    >
      <div
        class="card"
      >
        <div
          class="card-body text-center"
        >
          <div
            class="mb-2"
          >
            <span
              class="avatar"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M399 384.2C376.9 345.8 335.4 320 288 320l-64 0c-47.4 0-88.9 25.8-111 64.2c35.2 39.2 86.2 63.8 143 63.8s107.8-24.7 143-63.8zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zm256 16a72 72 0 1 0 0-144 72 72 0 1 0 0 144z"
                />
              </svg>
            </span>
          </div>
          <h5
            class="card-title mb-0"
          >
            mtsamis
          </h5>
          <p
            class="card-text text-muted"
          >
            Mike Tsamis
          </p>
          <hr />
          <div
            class="row text-center"
          >
            <div
              class="col-6"
            >
              <h6
                class="mb-0"
              >
                5
              </h6>
              <small
                class="text-muted"
              >
                Total Score
              </small>
            </div>
            <div
              class="col-6"
            >
              <h6
                class="mb-0"
              >
                2
              </h6>
              <small
                class="text-muted"
              >
                Ranking
              </small>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
